changefeedccl: fix cloudstorage filename ordering

asg0451 opened the pull request:
This change both introduces a versioning scheme for cloudstorage filenames
to allow for schema evolution without breaking ordering and fixes a bug
where filenames with the same timestamp were not ordered correctly.

The file naming version is now the first element
of the `uniquer` component of the name. It looks
like: `<timestamp>-$001-...rest`.

Epic: None

Release note (bug fix): Added versioning to cloudstorage
filenames to allow for schema evolution without
breaking ordering. Fixed a bug where filenames
with the same timestamp were not ordered
correctly under some circumstances.

asg0451 committed:
changefeedccl: add unit test for cloud storage sink ordering issue

Background: the cloud storage sink writes files to
cloud storage with names whose first component is
a timestamp. Beyond the normal changefeed ordering
guarantees, there is an additional guarantee that
if you read these files in lexicographic order,
the observed events will respect the changefeed
ordering guarantee.

However, under certain conditions (partially
lagging + restarts), we could emit 2 files with
the same timestamp name component that, if read in
order, resulted in a violation of the changefeed
ordering guarantee.

This patch adds a regression test for
this specific issue.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/155529)
<!-- Reviewable:end -->

asg0451 commented:
TODO: the proposed fix won't work. we have to think of another way

asg0451 committed:
changefeedccl: fix cloudstorage filename ordering

This change both introduces a versioning scheme for cloudstorage filenames
to allow for schema evolution without breaking ordering and fixes a bug
where filenames with the same timestamp were not ordered correctly.

The file naming version is now the first element
of the `uniquer` component of the name. It looks
like: `<timestamp>-~001-...rest`. There is an additional timestamp
component added thereafter for ordering tiebreaking.

Epic: None

Release note (bug fix): Added versioning to cloudstorage
filenames to allow for schema evolution without
breaking ordering. Fixed a bug where filenames
with the same timestamp were not ordered
correctly under some circumstances.

aerfrei commented on pkg/ccl/changefeedccl/sink_cloudstorage.go:
```diff
@@ -128,6 +129,10 @@ func (f *cloudStorageSinkFile) Write(p []byte) (int, error) {
 
+// filenameVersion is the current version of the filename scheme. This must be
+// incremented when the filename scheme is changed.
+const filenameVersion = 1
```
I'm thinking we're not expecting to have to have many versions of this, so for example, we wouldn't have to worry about the possibility of having a version 10+. Just confirming that that's right.

aerfrei reviewed (commented):
I'm still looking over this but left some initial comments.

asg0451 commented on pkg/ccl/changefeedccl/sink_cloudstorage_test.go:
the session id is random, so setting it here just makes the test less flaky

asg0451 commented on pkg/ccl/changefeedccl/sink_cloudstorage.go:
```diff
@@ -216,10 +221,14 @@ func (f *cloudStorageSinkFile) Write(p []byte) (int, error) {
-// cloudStorageSink in a running process, `<file_id>` is a unique id for each file written
-// by a given `<sink_id>` and <session_id> is a unique identifying string for the job
-// session running the `changeAggregator` that owns this sink.
+// `$<version_id>-<now_timestamp>-<session_id>-<node_id>-<sink_id>-<file_id>` where
```
yep, during the cutover

aerfrei replied:
I'm seeing this go in the opposite direction than I expected.
```
$123 < 123 true
$abc < abc true
```

Should we use something like ~ instead?

aerfrei commented on pkg/ccl/changefeedccl/sink_cloudstorage_test.go:
Do we want to merge this with the file contents debugging dump?

aerfrei commented on pkg/ccl/changefeedccl/sink_cloudstorage.go:
```diff
@@ -807,15 +831,22 @@ func (s *cloudStorageSink) flushFile(ctx context.Context, file *cloudStorageSink
-		s.jobSessionID, s.srcID, s.sinkID, fileID, file.topic, file.schemaID, s.ext)
-	if s.prevFilename != "" && filename < s.prevFilename {
+	var uniquer string
+	if s.includeUniqueTSInFilename {
```
My sense is that we can't safely turn this off since, if it wrote any files, they would all be lexicographically after the files that we'd start writing once we turn it off again. If so, should we not put it behind a cluster setting like this?

asg0451 replied:
if we're trying to backport this, probably the cluster setting needs to guard the entire file name format change.
